directives/animate/animate.js

```diff
@@ -0,0 +1,250 @@
+import { directive, PropertyPart } from 'lit-html';
+import { getComposedActiveElement, getNextFocusable } from '../../helpers/focus.js';
+import { getComposedParent, isComposedAncestor } from '../../helpers/dom.js';
+
+const stateMap = new WeakMap();
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+const showTransitionDuration = 300;
+const hideTransitionDuration = 200;
+const moveYValue = 20;
+
+class AnimationState {
+
+	constructor(propertyPart) {
+
+		if (!(propertyPart instanceof PropertyPart) || propertyPart.committer.name !== 'animate') {
+			throw new Error('animation directives must be used with "animation" property');
+		}
+
+		this.id = 0;
+		this.clone = null;
+		this.elem = propertyPart.committer.element;
+		this.state = 'unknown';
+
+	}
+
+	async animate(info) {
+
+		if (this.clone === null) {
+			this.clone = document.createElement('div');
+		}
+		this.elem.parentNode.insertBefore(this.clone, this.elem);
+
+		const id = ++this.id;
+
+		let outerResolve;
+		const onTransitionEnd = (event) => {
+			if (event.target === this.clone) {
+				this.clone.removeEventListener('transitionend', onTransitionEnd);
+				if (this.id === id) {
+					this.clone.remove();
+					this.clone = null;
+					info.onTransitionEnd();
+					['height', 'left', 'opacity', 'position', 'top', 'transform', 'transition', 'width']
+						.forEach((prop) => this.elem.style.removeProperty(prop));
+					outerResolve();
+				}
+			}
+		};
+		this.clone.addEventListener('transitionend', onTransitionEnd);
+
+		Object.assign(this.clone.style, info.clone.start);
+		Object.assign(this.elem.style, info.elem.start);
+
+		if (info.elem.transition) {
+			await new Promise((r) => requestAnimationFrame(r));
+			Object.assign(this.elem.style, info.elem.transition);
+		}
+
+		await new Promise((r) => requestAnimationFrame(r));
+		Object.assign(this.clone.style, info.clone.end);
+		Object.assign(this.elem.style, info.elem.end);
+
+		return new Promise((resolve) => outerResolve = resolve);
+
+	}
+
+	async getElemInfo() {
+
+		await new Promise((r) => requestAnimationFrame(r));
+		const style = window.getComputedStyle(this.elem);
+
+		if (style.display !== 'flex') {
+			this.elem.style.display = 'grid'; // forces margins to collapse during size calculation
+		}
+
+		const rect = this.elem.getBoundingClientRect();
+
+		let top = 0;
+		let left = 0;
+		const offsetParent = this.elem.offsetParent;
+		let e = this.elem;
+		while (e && e !== document && !(e instanceof DocumentFragment)) {
```

Comment:
The purpose of this code is to travel up the DOM searching for the `offsetParent`. `DocumentFragment`s [have no parent](https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment), so if we find one we need to stop traveling up in search of the offsetParent. Our `getOffsetParent()` DOM helper does something similar. It will not stop for a shadowRoot.